src/Repository/DBALEventRepository.php

```diff
@@ -68,22 +83,65 @@ public function appendTo(string $streamName, AggregateRootId $aggregateRootId, a
     }
 
     /**
-     * @return Event[]
-     *
      * @throws DBALException
+     * @throws EncodingFailed
      */
-    public function load(string $streamName, AggregateRootId $aggregateRootId): array
+    public function load(string $streamName, AggregateRootId $aggregateRootId): Stream
     {
         $result = $this->connection->executeQuery(
             'SELECT * FROM '.$streamName.' WHERE aggregate_id = ? ORDER BY version ASC',
             [(string) $aggregateRootId]
         );
 
+        $metadata = $this->loadMetadata($streamName, $aggregateRootId);
+
         $events = [];
         while (false !== $eventData = $result->fetch()) {
-            $events[] = $this->eventConverter->createFromArray($eventData);
+            $events[] = $this->eventConverter->createFromArray(
+                $eventData['event_name'],
+                [
+                    'event_id' => $eventData['event_id'],
+                    'aggregate_id' => $eventData['aggregate_id'],
+                    'payload' => $this->jsonEncoder->decode($eventData['payload']),
+                    'metadata' => $this->jsonEncoder->decode($eventData['metadata']),
+                    'created_at' => $eventData['created_at'],
+                    'version' => $eventData['version'],
+                ],
+                $metadata
+            );
         }
 
-        return $events;
+        return new Stream($events, $metadata);
+    }
+
+    /**
+     * @throws DBALException
+     * @throws EncodingFailed
+     */
+    public function loadMetadata(string $streamName, AggregateRootId $aggregateRootId): StreamMetadata
```

Comment:
I would separate this to a separate repository, as this gives this class too much responsibility